File: LeetCode/LeetCode/simplify_path.go
package main

import "fmt"
import "os"
import "strings"

func simplifyPath(path string) string {
	lp := len(path)
	stack := make([]string, 0, lp/2);
	dir := make([]byte, 0, lp)

	for i := 0; i < lp; i++ {
		dir = dir[:0]
		for i < lp && path[i] != '/' {
			dir = append(dir, path[i])
			i++
		}

		s := string(dir)

		switch s {
		case ".", "":
			//do nothing
		case "..":
			if len(stack) > 0 {
				stack = stack[:len(stack)-1] //pop
			}
		default:
			stack = append(stack, s)
		}
	}
	return "/" + strings.Join(stack, "/")
}

func main() {
	fmt.Println(simplifyPath(os.Args[1]))
}

File: LeetCode/LeetCode/simplify_path.cc
#include<iostream>
#include<sstream>
#include<vector>
#include<string>
using namespace std;
class Solution{
    public:
        string simplifyPath(string path) {
            vector<string> names;
            vector<string> tokens(move(split(path, '/')));
            for(const auto & token : tokens) {
                if(token == ".." && !names.empty()) {
                    names.pop_back();
                }else if(token != ".." && token != "." && !token.empty()) {
                    names.emplace_back(token);
                }
            }
            return string("/").append(join(names,'/'));
        }

    private:
        vector<string> split(const string &s, const char delim) {
            vector<string> tokens;
            stringstream ss(s);
            string token;
            while(getline(ss, token, delim)) {
                tokens.emplace_back(token);
            }
            return tokens;
        }

        string join(const vector<string> &names, const char delim) {
            ostringstream ss;
            if(!names.empty()) {
                const string delim_str(1, delim);
                copy(names.cbegin(), prev(names.cend()),
                        ostream_iterator<string>(ss, delim_str.c_str()));
                ss << names.back();
            }
            return ss.str();
        }
};

int main(int argc, char** argv) 
{
    printf("%s\n", Solution().simplifyPath(argv[1]).c_str());
    return 0;
}


